tiled/server/authentication.py

```diff
@@ -233,6 +218,92 @@ async def get_session_state(decoded_access_token=Depends(get_decoded_access_toke
+    if ("api_key" in request.query_params) and (
+        request.cookies.get(API_KEY_COOKIE_NAME) != api_key
+    ):
+        request.state.cookies_to_set.append(
+            {"key": API_KEY_COOKIE_NAME, "value": api_key}
+        )
+
+
+async def get_scopes_from_api_key(
+    api_key: str, settings: Settings, authenticators, db
+) -> Sequence[str]:
+    if not authenticators:
+        # Tiled is in a "single user" mode with only one API key.
+        return (
+            {
+                "read:metadata",
+                "read:data",
+                "write:metadata",
+                "write:data",
+                "create",
+                "register",
+                "metrics",
+            }
+            if secrets.compare_digest(api_key, settings.single_user_api_key)
+            else set()
+        )
+    # Tiled is in a multi-user configuration with authentication providers.
+    # We store the hashed value of the API key secret.
+    # By comparing hashes we protect against timing attacks.
+    # By storing only the hash of the (high-entropy) secret
+    # we reduce the value of that an attacker can extracted from a
+    # stolen database backup.
+    try:
+        secret = bytes.fromhex(api_key)
+    except Exception:
+        return set()
+    api_key_orm = await lookup_valid_api_key(db, secret)
+    if api_key_orm is None:
+        return set()
+    else:
+        principal = api_key_orm.principal
+        principal_scopes = set().union(*[role.scopes for role in principal.roles])
+        # This intersection addresses the case where the Principal has
+        # lost a scope that they had when this key was created.
+        scopes = set(api_key_orm.scopes).intersection(principal_scopes | {"inherit"})
+        if "inherit" in scopes:
+            # The scope "inherit" is a metascope that confers all the
+            # scopes for the Principal associated with this API,
+            # resolved at access time.
+            scopes.update(principal_scopes)
+        return scopes
+
+
+async def get_current_scopes(
+    decoded_access_token: Optional[dict[str, Any]] = Depends(get_decoded_access_token),
+    api_key: Optional[str] = Depends(get_api_key),
+    settings: Settings = Depends(get_settings),
+    authenticators=Depends(get_authenticators),
+    db=Depends(get_database_session),
+) -> set[str]:
+    if api_key is not None:
+        return await get_scopes_from_api_key(api_key, settings, authenticators, db)
+    elif decoded_access_token is not None:
+        return decoded_access_token["scp"]
+    else:
+        return {"read:metadata", "read:data"} if settings.allow_anonymous_access else {}
+
+
+async def check_scopes(
+    request: Request,
+    security_scopes: SecurityScopes,
+    scopes: set[str] = Depends(get_current_scopes),
+) -> None:
+    if not set(security_scopes.scopes).issubset(scopes):
+        raise HTTPException(
+            status_code=HTTP_401_UNAUTHORIZED,
+            detail=(
```

Comment:
Maybe this should be a separate issue, but I've often wished that this message differentiated between an unauthenticated request ("Maybe you need to log in?") and an authenticated one (current message is fine).

Reply:
Leaving for a separate issue as it will change test behaviour.

Reply:
https://github.com/bluesky/tiled/issues/925